library/alloc/src/collections/btree/testing/crash_test.rs

```diff
@@ -0,0 +1,147 @@
+use crate::fmt::Debug;
+use std::cmp::Ordering;
+use std::sync::atomic::{AtomicU64, Ordering::SeqCst};
+
+/// A recording of particular events happening in a test scenario that involves
+/// crash test dummies, some of which may be configured to panic at some point.
+/// Events are `clone`, `drop` or some anonymous `query`.
+///
+/// Crash test dummies are identified and ordered by an id, so they can be used
+/// as keys in a BTreeMap. The implementation intentionally uses only primitives.
+#[derive(Debug)]
+pub struct CrashTest {
+    bits_per_id: usize,
+    cloned: AtomicU64,
+    dropped: AtomicU64,
+    queried: AtomicU64,
+}
+
+impl CrashTest {
+    /// Sets up the recording of a test scenario that counts how many events
+    /// dummies experience, per event type and per dummy id. The range of dummy
+    /// ids is limited to 16 and the number of times all dummies with the same
+    /// dummy id experience the same event is also limited to 16.
+    pub fn new() -> Self {
+        Self {
+            bits_per_id: 4,
+            cloned: AtomicU64::new(0),
+            dropped: AtomicU64::new(0),
+            queried: AtomicU64::new(0),
+        }
+    }
+
+    /// Sets up the recording of a test scenario that counts how many events
+    /// dummies experience, per event type. There is virtually no limit on the
+    /// range of dummy ids or on the number of events happening.
+    pub fn new_totaling() -> Self {
+        Self {
+            bits_per_id: 0,
+            cloned: AtomicU64::new(0),
+            dropped: AtomicU64::new(0),
+            queried: AtomicU64::new(0),
+        }
+    }
+
+    /// Creates a crash test dummy that records what events it experiences
+    /// and optionally panics.
+    pub fn dummy(&self, id: usize, panic: Panic) -> Dummy<'_> {
+        assert!(self.is_supported_id(id));
+        Dummy { id, panic, context: self }
+    }
+
+    /// Returns how many times a dummy has been cloned.  If set up by `new`,
+    /// this is a hexadecimal composition of the count for each dummy id.
+    /// If set up by `new_totaling`, this is the sum over all dummies.
+    pub fn cloned(&self) -> u64 {
+        self.cloned.load(SeqCst)
+    }
+
+    /// Returns how many times a dummy has been dropped. If set up by `new`,
+    /// this is a hexadecimal composition of the count for each dummy id.
+    /// If set up by `new_totaling`, this is the sum over all dummies.
+    pub fn dropped(&self) -> u64 {
+        self.dropped.load(SeqCst)
+    }
+
+    /// Returns how many times a dummy has been queried. If set up by `new`,
+    /// this is a hexadecimal composition of the count for each dummy id.
```

Comment:
Can we add an example or documentation on what a hexadecimal composition *means* to the CrashTest struct? It sounds like what this is doing is splitting the u64 into 16 "sections", each 4 bits large, which enables us to track the number for each of those sections (up to 15, I guess). But I'm a bit confused as to why we need something so complicated over a vector of atomics, for example, or even a `Mutex<Vec<u64>>` - I don't think we care about performance here?

Reply:
I don't see the complication, but I'm also familiar with binary coded decimal registers. Instead I think involving a mutex complicates matters. Involving a vector probably simplifies, but it is (another) dependency of the btree test code on a cousin in the crate - to some degree, testing the branch you're sitting on. Given the simple use case, I prefer to pull the counters apart in individual local variables and leave the mapping over to the compiler.

The way you comment "up to 15, I guess" suggest to me you didn't notice that I tried to document that same point in the `CrashTest::new` member, and neither of us noticed I incorrectly said it's limited to 16, so I'm strengthened in my belief there's already too much documentation here.

Anyway, I think the next iteration away from the original is substantially better.